exporters/dsexporter/dsexporter_test.go

```diff
@@ -2,45 +2,136 @@ package main
-	metricFamilies, err := metrics.Gather()
-	assert.NoError(t, err)
+    client := NewKubeClient()
+    assert.Nil(t, client, "Client is not nil")
+}
+
+func TestGetDataSourceAvailability(t *testing.T) {
+	datasources := []string{"prometheus-appstudio-ds", "prometheus-appstudio", "appstudio-ds"}
+	check := "prometheus-appstudio-ds"
+	result := GetDataSourceAvailability(datasources, check)
+	if result != 1 {
+		t.Errorf("Expected result 1, but got %f", result)
+	}
 
-	var RequestCountValue float64
-	for _, mf := range metricFamilies {
-		if mf.GetName() == "request_count" {
-			RequestCountValue = mf.GetMetric()[0].GetCounter().GetValue()
-			break
+	check = "prometheus-appstudio-dsexporter"
+	result = GetDataSourceAvailability(datasources, check)
+	if result != 0 {
+		t.Errorf("Expected result 0, but got %f", result)
+	}
+}
+
+func TestGetGrafanaResource(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/apis/grafana.integreatly.org/v1beta1/namespaces/appstudio-grafana/grafanas/grafana-oauth" {
+			t.Errorf("Unexpected request path: %s", r.URL.Path)
 		}
+		response := `{"test": "data"}`
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(response))
+	}))
+	defer server.Close()
+
+	config := &rest.Config{
+		Host: server.URL,
 	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if(err != nil) {
+		t.Fatalf("Error: %v", err)
+	}
+	result := GetGrafanaResource(clientset)
+	expectedResult := map[string]interface{}{"test": "data"}
+	if !reflect.DeepEqual(result, expectedResult) {
+		t.Errorf("Expected %v, but got %v", expectedResult, result)
+	}
+}
 
-	// Check whether the exported metric value is initially 0.
-	assert.Equal(t, float64(0), RequestCountValue)
+func TestGetDataSources(t *testing.T) {
+	grafanaRes := map[string]interface{}{
+		"status": map[string]interface{}{
+			"datasources": []interface{}{"appstudio-ds1", "appstudio-ds2"},
+		},
+	}
 
-	// Increment the requestCounter by calling the Inc method.
-	exporter.requestCounter.Inc()
+	expectedResult := []string{"appstudio-ds1", "appstudio-ds2"}
+	result := GetDataSources(grafanaRes)
+	if !reflect.DeepEqual(result, expectedResult) {
+		t.Errorf("Test-1 failed, Expected %v, but got %v", expectedResult, result)
+	}
 
-	// Collecting metrics again
-	metricFamilies, err = metrics.Gather()
-	assert.NoError(t, err)
+	// empty datasources
+	grafanaRes = map[string]interface{}{
+		"status": map[string]interface{}{
+			"datasources": []interface{}{},
+		},
+	}
 
-	for _, mf := range metricFamilies {
-		if mf.GetName() == "request_count" {
-			RequestCountValue = mf.GetMetric()[0].GetCounter().GetValue()
-			break
-		}
+	expectedResult = []string{}
+	result = GetDataSources(grafanaRes)
+	if !reflect.DeepEqual(result, expectedResult) {
+		t.Errorf("Test-2 failed, Expected %v, but got %v", expectedResult, result)
+	}
+}
+
+func TestMain(t *testing.T) {
```

Comment:
What are we testing here? Is our code being called?

Reply:
here the server URL is some mock URL but we have exporter defined here same as our actual code and checks whether it's setting the labels and availability value correctly.

Reply:
I think the tests should run our actual code. Testing a copy of the code inside a test does not ensure we're testing the same thing and it also does not protect us from regressions (which is the main purpose of having unit tests in the first place).

Suppose that we make a change that breaks the, the test will still pass as it runs the old code.

All in all, we need to write the code in such a way that will allow us to cover as much of it with unit tests as possible. main functions are sometime trickier to test, so we need to keep them as simple as possible.

Reply:
Updated the test case to call `NewCustomCollector` and run our code.